src/main/kotlin/org/jitsi/videobridge/transport/octo/OctoTransport.kt

```diff
@@ -0,0 +1,269 @@
+
+/**
+ * [OctoTransport] handles *all* incoming and outgoing Octo traffic for a
+ * given bridge.
+ *
+ * [relayId] is the Octo relay ID which will be advertised by this JVB.
+ * Other bridges can use this ID in order to discover the socket address that
+ * this bridge is accessible on.  With the current implementation the ID just
+ * encodes a pre-configured IP address and port, e.g. "10.0.0.1:20000"
+ */
+class OctoTransport(
+    val relayId: String,
+    parentLogger: Logger
+) {
+    private val logger = createChildLogger(parentLogger)
+
+    private val stats = Stats()
+
+    /**
+     * Handlers for incoming Octo packets.  Packets will be routed to a handler based on the conference
+     * ID in the Octo packet
+     */
+    private val incomingPacketHandlers: MutableMap<String, IncomingOctoPacketHandler> = ConcurrentHashMap()
+
+    /**
+     * Maps how many Octo packets have been received for unknown conference IDs, to avoid
+     * spamming the error logs
+     */
+    private val unknownConferences: MutableMap<String, AtomicLong> = ConcurrentHashMap()
+
+    /**
+     * The handler which will be invoked when this [OctoTransport] wants to
+     * send data.
+     */
+    var outgoingDataHandler: OutgoingOctoPacketHandler? = null
+
+    /**
+     * Registers a [PacketHandler] for the given [conferenceId]
+     */
+    fun addHandler(conferenceId: String, handler: IncomingOctoPacketHandler) {
+        logger.info("Adding handler for conference $conferenceId")
+
+        synchronized(incomingPacketHandlers) {
+            incomingPacketHandlers.put(conferenceId, handler)?.let {
+                logger.warn("Replacing an existing packet handler for gid=$conferenceId")
+            }
+            unknownConferences.remove(conferenceId)
+        }
+    }
+
+    /**
+     * Removes the [PacketHandler] for the given [conferenceId] IFF the one
+     * in the map is the same as [handler].
+     */
+    fun removeHandler(conferenceId: String, handler: IncomingOctoPacketHandler) = synchronized(incomingPacketHandlers) {
+        // If the Colibri conference for this GID was re-created, and the
+        // original Conference object is expired after a new packet handler
+        // was registered, the new packet handler should not be removed (as
+        // this would break the new conference).
+        incomingPacketHandlers[conferenceId]?.let {
+            if (it == handler) {
+                logger.info("Removing handler for conference $conferenceId")
+                incomingPacketHandlers.remove(conferenceId)
+            } else {
+                logger.info("Tried to remove handler for conference $conferenceId but it wasn't the currently active one")
+            }
+        }
+    }
+
+    fun stop() {
+    }
+
+    fun dataReceived(buf: ByteArray, off: Int, len: Int, receivedTime: Instant) {
+        val (conferenceId, mediaType, sourceEndpointId) = try {
+            Triple(
+                OctoPacket.readConferenceId(buf, off, len),
```

Comment:
Is this efficient? Since it happens for each octo packt received, even something like allocating a `Triple` could have an effect (probably small, but still).

Reply:
I changed it...it feels painfully un-kotlin, and I'm tempted to say the `Triple` allocation won't be a big deal, but I'm not sure I'll get around to profiling it one way vs another.